test/broadway_test.exs

```diff
@@ -1760,6 +1764,157 @@ defmodule BroadwayTest do
+      end
+
+      {:ok, _broadway} =
+        Broadway.start_link(CustomHandlers,
+          name: broadway_name,
+          producer: [
+            module: {ManualProducer, %{test_pid: test_pid}},
+            stages: 1,
+            # We use a long rate limiting interval so that we can trigger the rate limiting
+            # manually from this test.
+            rate_limiting: [allowed_messages: 2, interval: 10000]
+          ],
+          processors: [default: [stages: 1, max_demand: 2]],
+          context: %{handle_message: handle_message}
+        )
+
+      send(
+        get_producer(broadway_name),
+        {:push_messages,
+         [
+           %Message{data: 1, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}},
+           %Message{data: 2, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}},
+           %Message{data: 3, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}}
+         ]}
+      )
+
+      # First, we assert that handle_demand is called and that two events are emitted.
+      assert_receive {:handle_demand_called, _demand}
+      assert_receive {:handle_message_called, %Message{data: 1}}
+      assert_receive {:handle_message_called, %Message{data: 2}}
+
+      # Now, since the rate limiting interval is long, we can assert that
+      # handle_demand is not called and that the third message is not emitted yet.
+      refute_receive {:handle_demand_called, _demand}
+      refute_receive {:handle_message_called, %Message{data: 3}}, 500
+
+      # We "cheat" and manually tell the rate limiter to reset the limit.
+      send(get_rate_limiter(broadway_name), :reset_limit)
+
+      assert_receive {:handle_demand_called, _demand}
+      assert_receive {:handle_message_called, %Message{data: 3}}
+    end
+
+    test "works when stopping a pipeline and draining producers" do
+      Process.flag(:trap_exit, true)
+
+      broadway_name = new_unique_name()
+      test_pid = self()
+
+      handle_message = fn message, _ ->
+        send(test_pid, {:handle_message_called, message})
+        message
+      end
+
+      {:ok, broadway} =
+        Broadway.start_link(CustomHandlers,
+          name: broadway_name,
+          producer: [
+            module: {ManualProducer, %{test_pid: test_pid}},
+            stages: 1,
+            # We use a long rate limiting interval so that we can trigger the rate limiting
+            # manually from this test.
+            rate_limiting: [allowed_messages: 2, interval: 10000]
+          ],
+          processors: [default: [stages: 1, max_demand: 2]],
+          context: %{handle_message: handle_message}
+        )
+
+      # First, we use all the rate limiting we have available.
+      send(
+        get_producer(broadway_name),
+        {:push_messages,
+         [
+           %Message{data: 1, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}},
+           %Message{data: 2, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}}
+         ]}
```

Comment:
In this test it is important to test that the buffer is made empty before we cancel. So I would have a third message here and make sure that it is delivered before `:message_during_cancel` (by adding the timestamps to those events, as we mentioned above).

Reply:
I have unresolved this comment because we didn't add the timestamps yet. We should also add the timestamps to the previous test, as mentioned here: https://github.com/plataformatec/broadway/pull/134

Reply:
I pushed a fix for this, but I don't know if the assertions are what you had in mind. :)

Reply:
Looks good. Let's just change the stage to use System.monotonic_time instead of System.system_time as this test could fail on systems with low time resolutions.